lib/srv/desktop/rdp/rdpclient/src/scard.rs

```diff
@@ -0,0 +1,1337 @@
+    SCARD_E_NO_ACCESS = 0x80100027,
+    SCARD_E_WRITE_TOO_MANY = 0x80100028,
+    SCARD_E_BAD_SEEK = 0x80100029,
+    SCARD_E_INVALID_CHV = 0x8010002A,
+    SCARD_E_UNKNOWN_RES_MSG = 0x8010002B,
+    SCARD_E_NO_SUCH_CERTIFICATE = 0x8010002C,
+    SCARD_E_CERTIFICATE_UNAVAILABLE = 0x8010002D,
+    SCARD_E_NO_READERS_AVAILABLE = 0x8010002E,
+    SCARD_E_COMM_DATA_LOST = 0x8010002F,
+    SCARD_E_NO_KEY_CONTAINER = 0x80100030,
+    SCARD_E_SERVER_TOO_BUSY = 0x80100031,
+    SCARD_E_PIN_CACHE_EXPIRED = 0x80100032,
+    SCARD_E_NO_PIN_CACHE = 0x80100033,
+    SCARD_E_READ_ONLY_CARD = 0x80100034,
+    SCARD_W_UNSUPPORTED_CARD = 0x80100065,
+    SCARD_W_UNRESPONSIVE_CARD = 0x80100066,
+    SCARD_W_UNPOWERED_CARD = 0x80100067,
+    SCARD_W_RESET_CARD = 0x80100068,
+    SCARD_W_REMOVED_CARD = 0x80100069,
+    SCARD_W_SECURITY_VIOLATION = 0x8010006A,
+    SCARD_W_WRONG_CHV = 0x8010006B,
+    SCARD_W_CHV_BLOCKED = 0x8010006C,
+    SCARD_W_EOF = 0x8010006D,
+    SCARD_W_CANCELLED_BY_USER = 0x8010006E,
+    SCARD_W_CARD_NOT_AUTHENTICATED = 0x8010006F,
+    SCARD_W_CACHE_ITEM_NOT_FOUND = 0x80100070,
+    SCARD_W_CACHE_ITEM_STALE = 0x80100071,
+    SCARD_W_CACHE_ITEM_TOO_BIG = 0x80100072,
+}
+
+#[derive(Debug)]
+#[allow(non_camel_case_types)]
+struct Long_Return {
+    return_code: ReturnCode,
+}
+
+impl Long_Return {
+    fn new(return_code: ReturnCode) -> Self {
+        Self { return_code }
+    }
+    fn encode(&self) -> RdpResult<Vec<u8>> {
+        let mut w = vec![];
+        w.write_i32::<LittleEndian>(self.return_code.to_i32().unwrap())?;
+        Ok(encode_response(encode_struct(w)?)?)
+    }
+}
+
+#[derive(Debug)]
+#[allow(non_camel_case_types)]
+struct EstablishContext_Call {
+    scope: Scope,
+}
+
+impl EstablishContext_Call {
+    fn decode(payload: &mut Payload) -> RdpResult<Self> {
+        let _header = RPCEStreamHeader::decode(payload)?;
+        let _header = RPCETypeHeader::decode(payload)?;
+        let scope = payload.read_u32::<LittleEndian>()?;
+        Ok(Self {
+            scope: Scope::from_u32(scope).ok_or(invalid_data_error(&format!(
+                "invalid smart card scope {:?}",
+                scope
+            )))?,
+        })
+    }
+}
+
+#[derive(Debug, FromPrimitive, ToPrimitive)]
+#[allow(non_camel_case_types)]
+enum Scope {
+    SCARD_SCOPE_USER = 0x00000000,
+    SCARD_SCOPE_TERMINAL = 0x00000001,
+    SCARD_SCOPE_SYSTEM = 0x00000002,
+}
+
+#[derive(Debug)]
+#[allow(non_camel_case_types)]
+struct EstablishContext_Return {
+    return_code: ReturnCode,
+    context: Context,
+}
+
+impl EstablishContext_Return {
+    fn new(return_code: ReturnCode, context: Context) -> Self {
+        Self {
+            return_code,
+            context,
+        }
+    }
+    fn encode(&self) -> RdpResult<Vec<u8>> {
+        let mut w = vec![];
+        w.write_i32::<LittleEndian>(self.return_code.to_i32().unwrap())?;
+        let mut index = 0;
+        w.extend_from_slice(&self.context.encode_ptr(&mut index)?);
```

Comment:
Can the encode functions take a buffer to write to instead of allocating their own? Generally I'd encourage all encode functions that don't produce a final message to take a `&dyn Write` or a `&mut Vec<u8>` instead of creating their own Vecs. Currently you're creating new Vecs in every nook and cranny which results in a bunch of wasted allocations. Coming from Go it's not always obvious but allocations really do harm performance. Strive to minimize them and instead pass writable buffers around, preferably preallocated to some expected/minimum size to avoid constant resizing.

Reply:
Yeah, that's true I guess. You'd essentially need a framework where every type implements some `encode` trait which also has a `size()` method which recursively calculates the size in advance.

Reply:
It's fine to leave this out for now though as it's a pretty big encoder refactor.